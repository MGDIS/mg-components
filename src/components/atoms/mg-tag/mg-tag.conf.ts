/**
 * List of all possibles variants
 */
export const variants = ['primary', 'secondary', 'success', 'warning', 'danger', 'info'] as const;

/**
 * TagVariantType type from tag variants
 */
export type TagVariantType = (typeof variants)[number];
