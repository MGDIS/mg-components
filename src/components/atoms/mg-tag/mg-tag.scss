/* Variables */
$mg-tag-border-width: 0.1rem;
$mg-tag-padding-horizontal: 0.7rem;

/* Mixins */
@mixin tag-variant($name, $backgroundColor, $color) {
  &.mg-tag--#{$name} {
    // Decoration
    background-color: hsl(var(--color-#{$backgroundColor}));
    border-color: hsl(var(--color-#{$backgroundColor}));
    // Font
    color: hsl(var(--color-#{$color}));

    &.mg-tag--outline {
      // Decoration
      background-color: transparent;
      // Font
      color: hsl(var(--color-#{$backgroundColor}));
    }
  }
}

/* Styles */
.mg-tag {
  // Display
  padding: calc((var(--mg-tag-height) - (2 * $mg-tag-border-width) - (var(--mg-tag-font-size) * var(--line-height))) / 2) $mg-tag-padding-horizontal - $mg-tag-border-width;
  min-height: var(--mg-tag-height);
  box-sizing: border-box;
  // Decoration
  border-radius: var(--mg-tag-border-radius);
  border: $mg-tag-border-width solid;
  // Font
  font-size: var(--mg-tag-font-size);
  font-weight: 600;
  text-align: center;

  // Variant
  @include tag-variant('primary', 'dark', 'light');
  @include tag-variant('secondary', 'light', 'dark');
  @include tag-variant('success', 'success', 'dark');
  @include tag-variant('warning', 'warning', 'dark');
  @include tag-variant('danger', 'danger', 'light');
  @include tag-variant('info', 'info', 'light');
}

.mg-tag,
mg-tag {
  display: inline-block;
  vertical-align: middle;
}
