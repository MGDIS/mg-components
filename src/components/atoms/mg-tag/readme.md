# mg-tag



<!-- Auto Generated Below -->


## Properties

| Property  | Attribute | Description           | Type     | Default            |
| --------- | --------- | --------------------- | -------- | ------------------ |
| `variant` | `variant` | Define button variant | `string` | `this.variants[0]` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
