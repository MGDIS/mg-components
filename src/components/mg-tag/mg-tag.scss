@mixin tag-variant($variant) {
  background-color: var(--tag-#{$variant}-background-color);
  color: var(--tag-#{$variant}-font-color);
}

@mixin tag-variants($variants...) {
  @each $variant in $variants {
    &.mg-tag--#{$variant} {
      @include tag-variant($variant);
    }
  }
}

.mg-tag {
  // Display
  padding: 0.3rem 0.7rem;
  // Decoration
  border-radius: 0.5rem;
  border: none;
  // Font
  font-family: 'Open sans';
  font-size: 1.2rem;
  font-weight: 600;
  line-height: 1.8rem;

  @include tag-variants('product', 'success', 'warning', 'danger', 'draft');
}
