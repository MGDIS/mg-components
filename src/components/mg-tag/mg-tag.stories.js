export default {
  name: 'blu',
  component: 'mg-tag',
  title: 'Atoms/mg-tag',
  parameters: {
    docs: {
      source: {
        code: `<mg-tag variant="brand">Label</mg-tag>`,
      },
    },
  },
  argTypes: {
    variant: {
      options: ['product', 'success', 'warning', 'danger', 'draft'],
      control: { type: 'select' },
    },
  },
};

const Template = args => <mg-tag {...args}>{args.slot}</mg-tag>;

export const MgInputText = Template.bind({});
MgInputText.args = {
  slot: 'Label',
  variant: 'product',
};
