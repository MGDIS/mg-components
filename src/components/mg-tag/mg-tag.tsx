import { Component, h, Prop, Watch } from '@stencil/core';

@Component({
  tag: 'mg-tag',
  styleUrl: 'mg-tag.scss',
  shadow: true,
})
export class MgTag {

  /**
   * Internal
   */
  private classes = ['mg-tag'];
  private variants = ["product" , "success" , "warning", "danger",  "draft"];

  /**
   * Define button variant
   */
   @Prop() variant: string = this.variants[0];
   @Watch('variant')
   validateVariant(newValue: string) {
     if(!this.variants.includes(newValue)) {
       throw new Error(`<mg-tag> prop "variant" must be one of : ${this.variants.join(', ')}`);
     }
     this.classes.push(`mg-tag--${this.variant}`);
   }

  /**
   * Check if props are well configured on init
   */
  componentWillLoad() {
    this.validateVariant(this.variant);
  }

  /**
   * Render
   */

  render() {
    return (
        <span class={this.classes.join(' ')}><slot></slot></span>
    );
  }

}
