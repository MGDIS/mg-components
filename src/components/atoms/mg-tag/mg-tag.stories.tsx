import { h } from '@stencil/core';
import { variants } from './mg-tag.conf';

export default {
  component: 'mg-tag',
  title: 'Atoms/mg-tag',
  argTypes: {
    variant: {
      options: variants,
      control: { type: 'select' },
      table: {
        defaultValue: { summary: variants[0] },
      },
    },
  },
};

const Template = args => {
  // Extract slot so it won't be render as an attribute
  const slot = args.slot;
  delete args.slot;
  // return element
  return <mg-tag {...args}>{slot}</mg-tag>;
};

export const MgTag = Template.bind({});
MgTag.args = {
  slot: 'Label',
  variant: variants[0],
  outline: false,
};
