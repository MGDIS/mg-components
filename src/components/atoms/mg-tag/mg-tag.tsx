import { Component, h, Prop, State, Watch } from '@stencil/core';
import { variants } from './mg-tag.conf';
import { ClassList } from '../../../utils/utils';
@Component({
  tag: 'mg-tag',
  styleUrl: 'mg-tag.scss',
  shadow: true,
})
export class MgTag {

  /**
   * Define button variant
   */
  @Prop() variant?: string = variants[0];
  @Watch('variant')
  validateVariant(newValue: string) {
    if(!variants.includes(newValue)) {
      throw new Error(`<mg-tag> prop "variant" must be one of : ${variants.join(', ')}.`);
    }
    this.classes.add(`mg-tag--${this.variant}`);
  }

  /**
   * Define if button is using outline style
   */
   @Prop() outline?: boolean;
   @Watch('outline')
   validateOutline(newValue: boolean) {
     if(newValue) this.classes.add(`mg-tag--outline`);
   }

  /**
   * Component classes
   */
  @State() classes: ClassList = new ClassList(['mg-tag']);

  /**
   * Check if props are well configured on init
   */
  componentWillLoad() {
    this.validateVariant(this.variant);
    this.validateOutline(this.outline);
  }

  /**
   * Render
   */

  render() {
    return (
        <span class={this.classes.join()}><slot></slot></span>
    );
  }

}
