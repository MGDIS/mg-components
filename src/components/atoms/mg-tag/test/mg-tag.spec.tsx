import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgTag } from '../mg-tag';

const getPage = (args, slot) => newSpecPage({
  components: [MgTag],
  template: () => (<mg-tag {...args}>{slot}</mg-tag>)
});

describe('mg-tag', () => {
  test.each(["product" , "success" , "warning", "danger",  "draft"])('Should render a %s tag', async (variant) => {
    const { root } = await getPage({variant}, variant);
    expect(root).toMatchSnapshot();
  });

  test.each(["", "blu", undefined])('Should throw error', async (variant) => {
    try {
      await getPage({variant}, "wrong variant");
    }
    catch (err){
      expect(err.message).toContain('<mg-tag> prop "variant" must be one of : ')
    }
  });
});
