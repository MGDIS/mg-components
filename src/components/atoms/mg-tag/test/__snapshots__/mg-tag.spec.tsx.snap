// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-tag Should render a danger tag 1`] = `
<mg-tag>
  <mock:shadow-root>
    <span class="mg-tag mg-tag--danger">
      <slot></slot>
    </span>
  </mock:shadow-root>
  danger
</mg-tag>
`;

exports[`mg-tag Should render a draft tag 1`] = `
<mg-tag>
  <mock:shadow-root>
    <span class="mg-tag mg-tag--draft">
      <slot></slot>
    </span>
  </mock:shadow-root>
  draft
</mg-tag>
`;

exports[`mg-tag Should render a product tag 1`] = `
<mg-tag>
  <mock:shadow-root>
    <span class="mg-tag mg-tag--product">
      <slot></slot>
    </span>
  </mock:shadow-root>
  product
</mg-tag>
`;

exports[`mg-tag Should render a success tag 1`] = `
<mg-tag>
  <mock:shadow-root>
    <span class="mg-tag mg-tag--success">
      <slot></slot>
    </span>
  </mock:shadow-root>
  success
</mg-tag>
`;

exports[`mg-tag Should render a warning tag 1`] = `
<mg-tag>
  <mock:shadow-root>
    <span class="mg-tag mg-tag--warning">
      <slot></slot>
    </span>
  </mock:shadow-root>
  warning
</mg-tag>
`;
